# **Pertemuan #9 - Pandas Data Structure**

Data Wrangling - Sekolah Data - Pacmann Academy

**Outline**

1. Review Topik
2. Study case 1: Mengolah nilai mahasiswa
2. Study case 2: Data pelanggan mall
2. Study case 3: Menggabungkan data pasien rumah sakit dari file berbeda
2. Study case 4: Mengakses data dari database

Hal-hal yang harus disiapkan sebelum menggunakan notebook ini.
- Upload file csv pada workspace yang dibutuhkan untuk notebook ini:
  - Mall_customer.csv
  - patient_A1.csv
  - patient_A2.csv
  - patient.xlsx
  - chinook.db
- import library pandas, yaitu library yang akan digunakan dalam notebook ini.

In [1]:
# import pandas 
import pandas as pd

## **Review Topik**
---

Dari materi yang sudah dipelajari di video:

### **Pandas dan Data Structure**

- Pandas package
- Series
- DataFrame

### **Membaca Data**

- Read data menggunakan fungsi read_csv()
- Parameter pada read_csv()
- Fungsi untuk membaca file data lain seperti sql, excel, stata

### **Casting type data**

- Mengubah type data pada series atau dataframe

### **Mengombinasikan data**

- Menggabung data menggunakan fungsi concat()
- Menggabung data menggunakan fungsi merge()
- Menggabung data menggunakan fungsi join()

### **Mengoksport data**

- Mengeksport dataframe menjadi file

## **Study Case 1: Mengolah nilai mahasiswa**
___

Anda mendapat tugas untuk mengolah data nilai mahasiswa di bawah ini.

|Nama|Nilai|
|-|-|
|Juchiro|83.54|
|Fandy|86.79|
|Erlangga|88.57|
|Matheos|71.48|
|Sugeng|72.65|
|Atma|92.73|
|Muliana|83.22|
|Iyya|77.72|
|Budiarto|85.09|
|Afendy|88.98|
|Muhaimin|68.40|

### 1.1
Buat series dari nilai tersebut. Nama sebagai index, nilai sebagai value


In [2]:
# buat data nilai dalam bentuk python dictionary
data_nilai = {
    "Juchiro": 83.54,
    "Fandy":86.79,
    "Erlangga":88.57,
    "Matheos":71.48,
    "Sugeng":72.65,
    "Atma":92.73,
    "Muliana":83.22,
    "Iyya":77.72,
    "Budiarto":85.09,
    "Afendy":88.98,
    "Muhaimin":68.40,
}

# buat series dari data dictionary
data = pd.Series(data_nilai)

# tampilkan data
data

Juchiro     83.54
Fandy       86.79
Erlangga    88.57
Matheos     71.48
Sugeng      72.65
Atma        92.73
Muliana     83.22
Iyya        77.72
Budiarto    85.09
Afendy      88.98
Muhaimin    68.40
dtype: float64

### 1.2
Tampilkan nilai Fandy dan Atma secara terpisah. Serta Tampilkan nilai Erlangga, Budiarto, dan Muhaimin bersamaan.

In [16]:
print('Nilai Fandy : ',data['Fandy'])
print('Nilai Atma : ',data['Atma'])
data[["Erlangga", "Budiarto", "Muhaimin"]]

Nilai Fandy :  86.79
Nilai Atma :  92.73


Erlangga    88.57
Budiarto    85.09
Muhaimin    68.40
dtype: float64

### 1.3
Tambahkan data berikut pada series yang telah dibuat

|Nama|Nilai|
|-|-|
|Yusril|86.35|
|Priyo|74.73|
|Yosua|79.12|

In [18]:
new_data = {
    "Yusril" : 86.35,
    "Priyo"  : 74.73,
    "Yosua"  : 79.12
}
n_series = pd.Series(new_data)
n_series

Yusril    86.35
Priyo     74.73
Yosua     79.12
dtype: float64

In [20]:
data = data.append(n_series)
data

Juchiro     83.54
Fandy       86.79
Erlangga    88.57
Matheos     71.48
Sugeng      72.65
Atma        92.73
Muliana     83.22
Iyya        77.72
Budiarto    85.09
Afendy      88.98
Muhaimin    68.40
Yusril      86.35
Priyo       74.73
Yosua       79.12
dtype: float64

### 1.4
Ubah nilai mahasiswa berikut ini:
* Muliana menjadi 92.45
* Iyya menjadi 79.80
* Juchiro dan Fandi masing-masing menjadi 90 dan 91 (secara bersamaan)

In [35]:
data["Muliana"]

92.45

In [33]:
data[["Muliana","Iyya","Juchiro","Fandy"]] = [92.45,79.80,90,91]

In [34]:
data

Juchiro     90.00
Fandy       91.00
Erlangga    88.57
Matheos     71.48
Sugeng      72.65
Atma        92.73
Muliana     92.45
Iyya        79.80
Budiarto    85.09
Afendy      88.98
Muhaimin    68.40
Yusril      86.35
Priyo       74.73
Yosua       79.12
dtype: float64

### 1.5
Export data menjadi file csv dengan nama "nilai_mahasiswa_update.csv"


In [41]:
data.to_csv("nilai_mahasiswa_update.csv")

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.
  """Entry point for launching an IPython kernel.


### Exercise

1. Ubah nilai Afendy dan Sugeng masing-masing menjadi 85.5 dan 77.77 
2. Ubah data nilai tersebut menjadi rentang nilai 1-10
3. Hapus nilai Matheos dan Fandy
4. Export data menjadi data excel

In [43]:
data[["Afendy","Sugeng"]] = [85.5,77.77]
data

Juchiro     90.00
Fandy       91.00
Erlangga    88.57
Matheos     71.48
Sugeng      77.77
Atma        92.73
Muliana     92.45
Iyya        79.80
Budiarto    85.09
Afendy      85.50
Muhaimin    68.40
Yusril      86.35
Priyo       74.73
Yosua       79.12
dtype: float64

In [46]:
data = data/10

In [47]:
data

Juchiro     9.000
Fandy       9.100
Erlangga    8.857
Matheos     7.148
Sugeng      7.777
Atma        9.273
Muliana     9.245
Iyya        7.980
Budiarto    8.509
Afendy      8.550
Muhaimin    6.840
Yusril      8.635
Priyo       7.473
Yosua       7.912
dtype: float64

In [49]:
data = data.drop(["Matheos","Fandy"])
data

Juchiro     9.000
Erlangga    8.857
Sugeng      7.777
Atma        9.273
Muliana     9.245
Iyya        7.980
Budiarto    8.509
Afendy      8.550
Muhaimin    6.840
Yusril      8.635
Priyo       7.473
Yosua       7.912
dtype: float64

In [55]:
data=data.round(2)

In [56]:
data

Juchiro     9.00
Erlangga    8.86
Sugeng      7.78
Atma        9.27
Muliana     9.25
Iyya        7.98
Budiarto    8.51
Afendy      8.55
Muhaimin    6.84
Yusril      8.64
Priyo       7.47
Yosua       7.91
dtype: float64

In [63]:
data.to_excel("nilai_mahasiswa_new.xlsx")

## **Study Case 2:  Data Pelanggan Mall**
___

Anda mendapat tugas sebagai admin dalam sebuah mall. Anda mendapat tugas untuk melakukan beberapa perubahan pada data pelanggan mall tersebut. Tapi, data tersebut tersimpan dalam bentuk file csv dengan nama "Mall_Customer.csv".

### 2.1
Buka data file csv tersebut dalam bentuk dataframe.

In [95]:
# Baca data csv menggunakan fungsi pada pandas.
data = pd.read_csv("Mall_Customers.csv")

# tampilkan data
data

CustomerID   Genre  Age  Annual_Income_(k$)  Spending_Score
0             1    Male   19                  15              39
1             2    Male   21                  15              81
2             3  Female   20                  16               6
3             4  Female   23                  16              77
4             5  Female   31                  17              40
..          ...     ...  ...                 ...             ...
195         196  Female   35                 120              79
196         197  Female   45                 126              28
197         198    Male   32                 126              74
198         199    Male   32                 137              18
199         200    Male   30                 137              83

[200 rows x 5 columns]

In [96]:
data.shape

(200, 5)

### 2.2
Index pada data merupakan angka dari 0 hingga 199. Anda diminta untuk mengubah Index menjadi angka hasil penjumlahan 1000 dengan "CustomerID". Misalnya CustomerID = 9, index menjadi 1009. Kemudian hapus  kolom "CustomerID"

In [97]:
idx = data["CustomerID"].values + 1000
data.index = idx
data

CustomerID   Genre  Age  Annual_Income_(k$)  Spending_Score
1001           1    Male   19                  15              39
1002           2    Male   21                  15              81
1003           3  Female   20                  16               6
1004           4  Female   23                  16              77
1005           5  Female   31                  17              40
...          ...     ...  ...                 ...             ...
1196         196  Female   35                 120              79
1197         197  Female   45                 126              28
1198         198    Male   32                 126              74
1199         199    Male   32                 137              18
1200         200    Male   30                 137              83

[200 rows x 5 columns]

In [98]:
data

CustomerID   Genre  Age  Annual_Income_(k$)  Spending_Score
1001           1    Male   19                  15              39
1002           2    Male   21                  15              81
1003           3  Female   20                  16               6
1004           4  Female   23                  16              77
1005           5  Female   31                  17              40
...          ...     ...  ...                 ...             ...
1196         196  Female   35                 120              79
1197         197  Female   45                 126              28
1198         198    Male   32                 126              74
1199         199    Male   32                 137              18
1200         200    Male   30                 137              83

[200 rows x 5 columns]

In [99]:
# Drop CustomerId
data.drop("CustomerID",axis=1, inplace=True)
data

Genre  Age  Annual_Income_(k$)  Spending_Score
1001    Male   19                  15              39
1002    Male   21                  15              81
1003  Female   20                  16               6
1004  Female   23                  16              77
1005  Female   31                  17              40
...      ...  ...                 ...             ...
1196  Female   35                 120              79
1197  Female   45                 126              28
1198    Male   32                 126              74
1199    Male   32                 137              18
1200    Male   30                 137              83

[200 rows x 4 columns]

### 2.3
Customer 1098 dan 1099 tidak lagi menjadi pelanggan pada mall tersebut.Anda diminta untuk menghapus data index 1098 dan 1099.

In [100]:
data = data.drop([1098,1099],axis=0)
data

Genre  Age  Annual_Income_(k$)  Spending_Score
1001    Male   19                  15              39
1002    Male   21                  15              81
1003  Female   20                  16               6
1004  Female   23                  16              77
1005  Female   31                  17              40
...      ...  ...                 ...             ...
1196  Female   35                 120              79
1197  Female   45                 126              28
1198    Male   32                 126              74
1199    Male   32                 137              18
1200    Male   30                 137              83

[198 rows x 4 columns]

### 2.4 
Ada data baru pada data customer. Anda Harus menambahkan data baru berikut.

||Genre|Age|Annual_Income_(k$)|Spending_Score
|-|-|-|-|-|
|1201|Male|22|20|83

In [101]:
n_df = pd.DataFrame({
    "Genre" : "Male",
    "Age"   : 22,
    "Annual_Income_(k$)":20,
    "Spending_Score":83},
    index=[1201]
)
n_df

Genre  Age  Annual_Income_(k$)  Spending_Score
1201  Male   22                  20              83

In [102]:
data = data.append(n_df)
data

Genre  Age  Annual_Income_(k$)  Spending_Score
1001    Male   19                  15              39
1002    Male   21                  15              81
1003  Female   20                  16               6
1004  Female   23                  16              77
1005  Female   31                  17              40
...      ...  ...                 ...             ...
1197  Female   45                 126              28
1198    Male   32                 126              74
1199    Male   32                 137              18
1200    Male   30                 137              83
1201    Male   22                  20              83

[199 rows x 4 columns]

### 2.5
Data income dan score kedepannya berpotensi berisi data float. Ubah type data income dan score menjadi float.

In [107]:
data['Annual_Income_(k$)'] = data['Annual_Income_(k$)'].astype(float)
data['Spending_Score'] = data["Spending_Score"].astype(float)
data

Genre  Age  Annual_Income_(k$)  Spending_Score
1001    Male   19                15.0            39.0
1002    Male   21                15.0            81.0
1003  Female   20                16.0             6.0
1004  Female   23                16.0            77.0
1005  Female   31                17.0            40.0
...      ...  ...                 ...             ...
1197  Female   45               126.0            28.0
1198    Male   32               126.0            74.0
1199    Male   32               137.0            18.0
1200    Male   30               137.0            83.0
1201    Male   22                20.0            83.0

[199 rows x 4 columns]

In [108]:
# tampilkan type data tiap kolom
data.dtypes

Genre                  object
Age                     int64
Annual_Income_(k$)    float64
Spending_Score        float64
dtype: object

### Exercise

1. Ubah masing-masing header sesuai data berikut:
  * Genre = JenisKelamin
  * Age = Umur
  * Annual_Income_(k$) = Pendapatan
  * Spending_Score = Skor
2. Tambah data berikut:

||JenisKelamin|Umur|Pendapatan|Skor
|-|-|-|-|-|
|1202|Male|23|100.5|50.4
|1205|Female|45|70.5|35.6
|1207|Male|35|88.3|54.8

3. Ubah umur data index 1200 menjadi 31 dan Update data index 1002 menjadi:

||JenisKelamin|Umur|Pendapatan|Skor
|-|-|-|-|-|
|1002|Male|22|20|83

4. Export data menjadi data excel

In [112]:
new_header = {
    "Genre":"JenisKelamin",
    "Age":"Umur",
    "Annual_Income_(k$)":"Pendapatan",
    "Spending_Score":"Skor"
}
data = data.rename(columns=new_header)
data

JenisKelamin  Umur  Pendapatan  Skor
1001         Male    19        15.0  39.0
1002         Male    21        15.0  81.0
1003       Female    20        16.0   6.0
1004       Female    23        16.0  77.0
1005       Female    31        17.0  40.0
...           ...   ...         ...   ...
1197       Female    45       126.0  28.0
1198         Male    32       126.0  74.0
1199         Male    32       137.0  18.0
1200         Male    30       137.0  83.0
1201         Male    22        20.0  83.0

[199 rows x 4 columns]

In [114]:
# Add new data
val = {
    "JenisKelamin" : ["Male","Female","Male"],
    "Umur": [23,45,35],
    "Pendapatan":[100.5,70.5,88.3],
    "Skor":[50.4,35.6,54.8]
}
idx = [1202,1205,1207]
n_df = pd.DataFrame(val,index=idx)
n_df

JenisKelamin  Umur  Pendapatan  Skor
1202         Male    23       100.5  50.4
1205       Female    45        70.5  35.6
1207         Male    35        88.3  54.8

In [116]:
data = data.append(n_df)
data

JenisKelamin  Umur  Pendapatan  Skor
1001         Male    19        15.0  39.0
1002         Male    21        15.0  81.0
1003       Female    20        16.0   6.0
1004       Female    23        16.0  77.0
1005       Female    31        17.0  40.0
...           ...   ...         ...   ...
1200         Male    30       137.0  83.0
1201         Male    22        20.0  83.0
1202         Male    23       100.5  50.4
1205       Female    45        70.5  35.6
1207         Male    35        88.3  54.8

[202 rows x 4 columns]

In [130]:
data.loc[[1200],["Umur"]] 

Umur
1200    31

In [127]:
# Ubah umur data index 1200 menjadi 31
data.loc[[1200],["Umur"]] = 31
data

JenisKelamin  Umur  Pendapatan  Skor
1001         Male    19        15.0  39.0
1002         Male    21        15.0  81.0
1003       Female    20        16.0   6.0
1004       Female    23        16.0  77.0
1005       Female    31        17.0  40.0
...           ...   ...         ...   ...
1200         Male    31       137.0  83.0
1201         Male    22        20.0  83.0
1202         Male    23       100.5  50.4
1205       Female    45        70.5  35.6
1207         Male    35        88.3  54.8

[202 rows x 4 columns]

In [129]:
# Update data index 1002
data.loc[1002] = ["Male",22,20,83]
data

JenisKelamin  Umur  Pendapatan  Skor
1001         Male    19        15.0  39.0
1002         Male    22        20.0  83.0
1003       Female    20        16.0   6.0
1004       Female    23        16.0  77.0
1005       Female    31        17.0  40.0
...           ...   ...         ...   ...
1200         Male    31       137.0  83.0
1201         Male    22        20.0  83.0
1202         Male    23       100.5  50.4
1205       Female    45        70.5  35.6
1207         Male    35        88.3  54.8

[202 rows x 4 columns]

In [132]:
data.to_excel("mall_customer_edited.xlsx")

## Study Case 3: Menggabungkan data pasien rumah sakit dari file berbeda.
Ada 4 file dari 2 rumah sakit yang berbeda yang harus anda gabung.
* File "patient_A1.csv" merupakan data pasien dari rumah sakit A yang berisi data Pregnancies, Glucose, BloodPressure, SkinThickness.
* File "patient_A2.csv" merupakan data pasien dari rumah sakit A yang berisi data Insulin, BMI, Diabetes, PedigreeFunction, Age, Outcome.
* File "patient.xlsx" merupakan data pasien dari rumah sakit B yang berisi data 
  - sheet patient_B1: IdPatient, Pregnancies, SkinThickness, BMI, Age, Outcome.
  - sheet patient_B2: IdPatient, Glucose, BloodPressure, DiabetesPedigreeFunction.

### 3.1
Buat dataframe dari setiap file csv yang diberikan.

In [134]:
# Baca data csv menggunakan fungsi pada pandas.
dataA1 = pd.read_csv("patient_A1.csv")

# tampilkan data
dataA1.head()

Pregnancies  Glucose  BloodPressure  SkinThickness
0            6      148             72             35
1            1       85             66             29
2            8      183             64              0
3            1       89             66             23
4            0      137             40             35

In [136]:
dataA2 = pd.read_csv("patient_A2.csv")
dataA2.head()

Insulin   BMI  DiabetesPedigreeFunction  Age  Outcome
0        0  33.6                     0.627   50        1
1        0  26.6                     0.351   31        0
2        0  23.3                     0.672   32        1
3       94  28.1                     0.167   21        0
4      168  43.1                     2.288   33        1

In [139]:
dataB1 = pd.read_excel("patient.xlsx",sheet_name="patient_B1")
dataB1.tail()

IdPatient  Pregnancies  SkinThickness   BMI  Age  Outcome
269        369           10             48  32.9   63        0
270        370            2             27  36.8   27        0
271        371            5             23  26.2   30        0
272        372            1              0  30.1   47        1
273        373            1             31  30.4   23        0

In [140]:
dataB2 = pd.read_excel("patient.xlsx",sheet_name="patient_B2")
dataB2.tail()

IdPatient  Glucose  BloodPressure  DiabetesPedigreeFunction
269        369      101             76                     0.171
270        370      122             70                     0.340
271        371      121             72                     0.245
272        372      126             60                     0.349
273        373       93             70                     0.315